clients/drcachesim/tracer/tracer.cpp

```diff
@@ -786,55 +934,21 @@ event_thread_init(void *drcontext)
     create_buffer(data);
 
     if (op_offline.get_value()) {
-        /* We do not need to call drx_init before using drx_open_unique_appid_file.
-         * Since we're now in a subdir we could make the name simpler but this
-         * seems nice and complete.
-         */
-        int i;
-        const int NUM_OF_TRIES = 10000;
-        uint flags = IF_UNIX(DR_FILE_CLOSE_ON_FORK |)
-            DR_FILE_ALLOW_LARGE | DR_FILE_WRITE_REQUIRE_NEW;
-        /* We use drx_open_unique_appid_file with DRX_FILE_SKIP_OPEN to get a
-         * file name for creation.  Retry if the same name file already exists.
-         * Abort if we fail too many times.
-         */
-        for (i = 0; i < NUM_OF_TRIES; i++) {
-            drx_open_unique_appid_file(logsubdir,
-                                       dr_get_thread_id(drcontext),
-                                       OUTFILE_PREFIX, OUTFILE_SUFFIX,
-                                       DRX_FILE_SKIP_OPEN,
-                                       buf, BUFFER_SIZE_ELEMENTS(buf));
-            NULL_TERMINATE_BUFFER(buf);
-            data->file = file_ops_func.open_file(buf, flags);
-            if (data->file != INVALID_FILE)
-                break;
-        }
-        if (i == NUM_OF_TRIES) {
-            NOTIFY(0, "Fatal error: failed to create trace file %s\n", buf);
-            dr_abort();
-        }
-        NOTIFY(2, "Created thread trace file %s\n", buf);
-
-        /* Write initial headers at the top of the first buffer. */
-        data->init_header_size =
-            instru->append_thread_header(data->buf_base, dr_get_thread_id(drcontext));
-        BUF_PTR(data->seg_base) = data->buf_base + data->init_header_size;
-        BUF_PTR(data->seg_base) +=
-            instru->append_tid(BUF_PTR(data->seg_base), dr_get_thread_id(drcontext));
-        BUF_PTR(data->seg_base) +=
-            instru->append_pid(BUF_PTR(data->seg_base), dr_get_process_id());
+        if (op_num_threads.get_value() == 0)
+            data->file = create_thread_file(dr_get_thread_id(drcontext));
+        else
+            data->file = INVALID_FILE;
     } else {
         /* pass pid and tid to the simulator to register current thread */
-        proc_info = (byte *)buf;
-        DR_ASSERT(BUFFER_SIZE_BYTES(buf) >= 3*instru->sizeof_entry());
-        proc_info += instru->append_thread_header(proc_info, dr_get_thread_id(drcontext));
-        proc_info += instru->append_tid(proc_info, dr_get_thread_id(drcontext));
-        proc_info += instru->append_pid(proc_info, dr_get_process_id());
-        write_trace_data(drcontext, (byte *)buf, proc_info);
-
-        /* put buf_base to TLS plus header slots as starting buf_ptr */
-        BUF_PTR(data->seg_base) = data->buf_base + buf_hdr_slots_size;
+        byte *buf = (byte *)
+            dr_raw_mem_alloc(max_buf_size, DR_MEMPROT_READ | DR_MEMPROT_WRITE, NULL);
```

Comment:
> What if there are thousands threads during attach?
> It may slowdown the attach, because the thread init event will be blocked by the enqueue.

This is a separate topic, right?  This is about the thread "pool" having just one thread for online and worries about it having worse performance than no pool at all?